### PR TITLE
GL-229: Add .loom support to Optimus

### PR DESCRIPTION
**Pull request by barkasn**
### Purpose

- Add loom support to Optimus

### Changes

- This PR adds loom support to the output of Optimus
- The ZARR output is now always generated and the options are changed accordingly
- Ticket GL-229

### Review Instructions

- No instructions.

### CODE REVIEW

File: docker/zarr-to-loom/optimus_zarr_to_loom.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+    del expr
+    expr_sp_t = sc.sparse.coo_matrix((expr_sp.data, (expr_sp.col, expr_sp.row)),
+                                     shape=(expr_sp.shape[1], expr_sp.shape[0]))
+    del expr_sp
+
+    # ROW/GENE Metadata
+
+    # Check that the first gene metadata column is the gene name as expected
+    if not (root["/pr_test_dataset.zarr/gene_metadata_string_name"][:] == 'gene_name')[0]:
+        raise UnexpectedInputFormat("The first gene metadata item is not the gene_name");
```
Comment (jessicaway) [inline docker/zarr-to-loom/optimus_zarr_to_loom.py]: You could do this check before the matrix manipulations to fail faster for bad input

File: docker/zarr-to-loom/optimus_zarr_to_loom.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+
+    # Prepare row attributes (gene attributes)
+    # Follow loom file Conventions
+    row_attrs = {
+        "Gene": root["/pr_test_dataset.zarr/gene_metadata_string"][:][0,],
```
Comment (jessicaway) [inline docker/zarr-to-loom/optimus_zarr_to_loom.py]: Does [0, ] work here?

File: docker/zarr-to-loom/optimus_zarr_to_loom.py [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+        sys.exit("Error: The output loom file exists!")
+
+    # Open the ZARR
+    store = zarr.DirectoryStore(input_zarr_path)
+    root = zarr.open(store)
+
+    # Get the expression matrix
+    # expression matrix in numpy ndarray format (dense)
+    # NOTE: If memory is limiting this could be done by chunk
+    expr = root["/pr_test_dataset.zarr/expression"][:]
```
Comment (jessicaway) [inline docker/zarr-to-loom/optimus_zarr_to_loom.py]: name is hardcoded here

File: docker/zarr-to-loom/optimus_zarr_to_loom.py [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+
+    numeric_field_names = root["/pr_test_dataset.zarr/gene_metadata_numeric_name"][:]
+
+    # Generate with a list comprehension
+    def add_to_gene_meta_by_index(i):
+        name = numeric_field_names[i]
+        data = root["/pr_test_dataset.zarr/gene_metadata_numeric"][:][:, i]
+        row_attrs[name] = data
+
+    [add_to_gene_meta_by_index(i) for i in range(0, numeric_field_names.shape[0])]
```
Comment (jessicaway) [inline docker/zarr-to-loom/optimus_zarr_to_loom.py]: use for loop for readability

File: library/tasks/ZarrUtils.wdl [hunk 5]
```diff
@@ -128,3 +128,45 @@ task OptimusZarrConversion {
...
+
+     Int preemptible = 3
+     Int cpu = 1
+
+     meta {
+         description: "This task converts the Optimus Zarr output into a loom file"
+     }
+
+     parameter_meta {
+         machine_mem_mb: "(optional) the amount of memory in (MiB) to provision for this task"
```
Comment (jessicaway) [inline library/tasks/ZarrUtils.wdl]: nit picky, but spacing looks slightly off here

File: library/tasks/ZarrUtils.wdl [hunk 6]
```diff
@@ -128,3 +128,45 @@ task OptimusZarrConversion {
...
+        cpu: 1
+        memory: "10 GiB"
+        disks: "local-disk 100 HDD"
+        preemptible: preemptible
+    }
+
+    output {
+       File loom_output = "output.loom"
+    }
+}
```
Comment (jessicaway) [inline library/tasks/ZarrUtils.wdl]: Add newline

File: pipelines/optimus/PublicDocumentation.md [hunk 7]
```diff
@@ -26,7 +26,7 @@ The bead-specific barcodes and UMIs are encoded on sequencing primers that also
 | Aligner           |STAR       |[Dobin, et al.,2013](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3530905/)|
 | Transcript Quantification |Utilities for processing large-scale single cell datasets |[Sctools](https://github.com/HumanCellAtlas/sctools) |                      
 |Data Input File Format |File format in which sequencing data is provided |[FASTQ](https://academic.oup.com/nar/article/38/6/1767/3112533) |                       
-|Data Output File Format |File formats in which Optimus output is provided |[BAM](http://samtools.github.io/hts-specs/), [Zarr version 2](https://zarr.readthedocs.io/en/stable/spec/v2.html), Python numpy arrays |
+|Data Output File Format |File formats in which Optimus output is provided |[BAM](http://samtools.github.io/hts-specs/), [Zarr version 2](https://zarr.readthedocs.io/en/stable/spec/v2.html), Python numpy arrays (internal), loom |
```
Comment (jessicaway) [inline pipelines/optimus/PublicDocumentation.md]: Should loom documentation also be linked here?

**Comment by kishorikonwar** [comment]
retest

File: docker/zarr-to-loom/optimus_zarr_to_loom.py [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+    pass
+
+
+def main():
+    # Parse the arguments
+    description = """This script converts ZARR optimus output to loom format"""
+    parser = argparse.ArgumentParser(description=description)
+    parser.add_argument('--input-zarr', dest="input_zarr_path", required=True, help="Path to input ZARR file")
+    parser.add_argument('--output-loom', dest="output_loom_path", required=True, help="Path to output loom file")
+    parser.add_argument('--sample_id', dest="sample_id", required=True, help="Sample identifier")
```
Comment (tlangs) [inline docker/zarr-to-loom/optimus_zarr_to_loom.py]: command-line arguments should consistently use `-` or `_` as separators

File: docker/zarr-to-loom/optimus_zarr_to_loom.py [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+        sys.exit("Error: The output loom file exists!")
+
+    # Open the ZARR
+    store = zarr.DirectoryStore(input_zarr_path)
+    root = zarr.open(store)
+
+    # Get the expression matrix
+    # expression matrix in numpy ndarray format (dense)
+    # NOTE: If memory is limiting this could be done by chunk
+    expr = root[f"/{sample_id}.zarr/expression"][:]
```
Comment (tlangs) [inline docker/zarr-to-loom/optimus_zarr_to_loom.py]: `/{sample_id}.zarr/` is repeated many times in the script. Might be worth it to pull it out as a constant.

File: library/tasks/ZarrUtils.wdl [hunk 10]
```diff
@@ -128,3 +128,46 @@ task OptimusZarrConversion {
...
+         machine_mem_mb: "(optional) the amount of memory in (MiB) to provision for this task"
+	 cpu: "(optional) the number of cpus to provision for this task"
+	 preemptible: "(optional) if non-zero, request a pre-emptible instance and allow for this number of preemptions before running the task on a non-preemptible machine"
+    }
+    
+    command {
+       set -euo pipefail
+
+       mkdir packed_zarr
+       echo ${sep=' ' zarr_files} > zarr_filenames.txt
```
Comment (tlangs) [inline library/tasks/ZarrUtils.wdl]: `zarr_filenames.txt` isn't used anywhere

File: pipelines/optimus/Optimus.wdl [hunk 11]
```diff
@@ -40,8 +40,8 @@ workflow Optimus {
   # this is used to scatter matched [r1_fastq, r2_fastq, i1_fastq] arrays
   Array[Int] indices = range(length(r1_fastq))
 
-  # whether to convert the outputs to Zarr format, by default it's set to true
-  Boolean output_zarr = true
+  # If true produce the optinal loom output
```
Comment (tlangs) [inline pipelines/optimus/Optimus.wdl]: "optional"

File: pipelines/optimus/Optimus.wdl [hunk 12]
```diff
@@ -208,33 +208,42 @@ workflow Optimus {
       col_index = MergeCountFiles.col_index
   }
 
-  if (output_zarr) {
-    call ZarrUtils.OptimusZarrConversion{
+  call ZarrUtils.OptimusZarrConversion{
+    input:
+      sample_id=sample_id,
+      annotation_file=annotations_gtf,
```
Comment (tlangs) [inline pipelines/optimus/Optimus.wdl]: There should be spaces surrounding the `=`.

File: pipelines/optimus/Optimus.wdl [hunk 13]
```diff
@@ -208,33 +208,42 @@ workflow Optimus {
...
+      sparse_count_matrix = MergeCountFiles.sparse_count_matrix,
+      cell_id = MergeCountFiles.row_index,
+      gene_id = MergeCountFiles.col_index,
+      empty_drops_result = RunEmptyDrops.empty_drops_result
+  }
+
+  if (output_loom) {
+    call ZarrUtils.OptimusZarrToLoom {
       input:
         sample_id=sample_id,
```
Comment (tlangs) [inline pipelines/optimus/Optimus.wdl]: There should be spaces surrounding the =.

File: test/optimus/pr/ValidateOptimus.wdl [hunk 14]
```diff
@@ -58,14 +69,14 @@ task ValidateBam {
 
         if [ "$calculated_checksum" == ${expected_checksum} ]
         then
-             echo Computed and expected bam hashes match \( $calculated_checksum \)
+             echo Computed and expected bam hashes match \( "$calculated_checksum" \)
              printf PASS > result.txt
         else 
-             echo Computed \( $calculated_checksum \) and expected \( ${expected_checksum} \) bam file hashes do not match
+             echo Computed \( "$calculated_checksum" \) and expected \( ${expected_checksum} \) bam file hashes do not match
```
Comment (tlangs) [inline test/optimus/pr/ValidateOptimus.wdl]: why is `calculated_checksum` quoted but `expected_checksum` not?

File: test/optimus/pr/ValidateOptimus.wdl [hunk 15]
```diff
@@ -143,16 +188,16 @@ task ValidateMetrics {
...
+            echo Computed \( "$gene_metric_hash" \) and expected \( ${expected_gene_metric_hash} \) gene checksums do not match
             fail=true
         fi
 
         if [ "$cell_metric_hash" == "${expected_cell_metric_hash}" ]; then
-            echo Computed and expected cell metrics match \( $cell_metric_hash \)
+            echo Computed and expected cell metrics match \( "$cell_metric_hash" \)
         else
-            echo Computed \( $cell_metric_hash \) and expected \( ${expected_cell_metric_hash} \) cell metrics hashes do not match
+            echo Computed \( "$cell_metric_hash" \) and expected \( ${expected_cell_metric_hash} \) cell metrics hashes do not match
```
Comment (tlangs) [inline test/optimus/pr/ValidateOptimus.wdl]: Why is `cell_metric_hash` quoted and `expected_cell_metric_hash` not?

File: library/tasks/ZarrUtils.wdl (hunk 5 above)
Comment (barkasn) [reply library/tasks/ZarrUtils.wdl]: done

File: library/tasks/ZarrUtils.wdl (hunk 6 above)
Comment (barkasn) [reply library/tasks/ZarrUtils.wdl]: done

File: pipelines/optimus/PublicDocumentation.md (hunk 7 above)
Comment (barkasn) [reply pipelines/optimus/PublicDocumentation.md]: done

File: test/optimus/pr/ValidateOptimus.wdl (hunk 14 above)
Comment (barkasn) [reply test/optimus/pr/ValidateOptimus.wdl]: done

File: pipelines/optimus/Optimus.wdl (hunk 11 above)
Comment (barkasn) [reply pipelines/optimus/Optimus.wdl]: done

File: docker/zarr-to-loom/optimus_zarr_to_loom.py (hunk 3 above)
Comment (barkasn) [reply docker/zarr-to-loom/optimus_zarr_to_loom.py]: fixed

File: docker/zarr-to-loom/optimus_zarr_to_loom.py (hunk 8 above)
Comment (barkasn) [reply docker/zarr-to-loom/optimus_zarr_to_loom.py]: done

File: library/tasks/ZarrUtils.wdl (hunk 10 above)
Comment (barkasn) [reply library/tasks/ZarrUtils.wdl]: good catch, was leftover from debug

File: pipelines/optimus/Optimus.wdl (hunk 12 above)
Comment (barkasn) [reply pipelines/optimus/Optimus.wdl]: done

File: test/optimus/pr/ValidateOptimus.wdl (hunk 15 above)
Comment (barkasn) [reply test/optimus/pr/ValidateOptimus.wdl]: done

File: docker/zarr-to-loom/optimus_zarr_to_loom.py (hunk 4 above)
Comment (barkasn) [reply docker/zarr-to-loom/optimus_zarr_to_loom.py]: done

File: pipelines/optimus/Optimus.wdl (hunk 13 above)
Comment (tlangs) [reply pipelines/optimus/Optimus.wdl]: This still needs spaces around the `=`

**Pull request merged**